pkg/controller/master/storagenetwork/storage_network.go

```diff
@@ -0,0 +1,518 @@
+		alertmanager:         alertmanager,
+		alertmanagerCache:    alertmanager.Cache(),
+		settings:             settings,
+		settingsCache:        settings.Cache(),
+		settingsController:   settings,
+		deployments:          deployments,
+		deploymentCache:      deployments.Cache(),
+		managedCharts:        managedCharts,
+		managedChartCache:    managedCharts.Cache(),
+	}
+
+	settings.OnChange(ctx, ControllerName, controller.OnStorageNetworkChange)
+	return nil
+}
+
+func (h *Handler) setConfiguredCondition(setting *harvesterv1.Setting, finish bool, reason string, msg string) (*harvesterv1.Setting, error) {
+
+	settingCopy := setting.DeepCopy()
+	if finish {
+		harvesterv1.SettingConfigured.True(settingCopy)
+	} else {
+		harvesterv1.SettingConfigured.False(settingCopy)
+	}
+
+	harvesterv1.SettingConfigured.Reason(settingCopy, reason)
+	harvesterv1.SettingConfigured.Message(settingCopy, msg)
+
+	return h.settings.Update(settingCopy)
+}
+
+// webhook needs check if VMs are off
+// webhook needs check if nad is existing
+func (h *Handler) OnStorageNetworkChange(key string, setting *harvesterv1.Setting) (*harvesterv1.Setting, error) {
+	if setting == nil || setting.DeletionTimestamp != nil || setting.Name != settings.StorageNetworkName {
+		return setting, nil
+	}
+
+	if ok, err := h.checkLonghornSetting(setting); ok {
+		// finish
+		return nil, nil
+	} else {
+		// reconcile
+		return nil, err
+	}
+
+	logrus.Infof("storage network change: %s", setting.Value)
+
+	// if replica eq 0, skip
+	// save replica to annotation
+	// set replica to 0
+	if !h.checkPodStatusAndStop() {
+		logrus.Infof("Requeue to check pod status again")
+		if _, err := h.setConfiguredCondition(setting, false, ReasonInProgress, MsgStopPod); err != nil {
+			logrus.Warnf("update status error %v", err)
+		}
+		h.settingsController.EnqueueAfter(setting.Name, 5*time.Second)
+		return nil, nil
+	}
+
+	logrus.Infof("all pods are stopped")
+
+	// check volume detach before put LH settings
+	if ok, err := h.checkLonghornVolumeDetach(); !ok {
+		if err != nil {
+			// log error only but not return error to controller
+			logrus.Warnf("check Longhorn volume error: %v", err)
+		}
+		logrus.Infof("still has attached volume")
+		if _, err := h.setConfiguredCondition(setting, false, ReasonInProgress, MsgWaitForVolumes); err != nil {
+			logrus.Warnf("update status error %v", err)
+		}
+		h.settingsController.EnqueueAfter(setting.Name, 5*time.Second)
+		return nil, nil
+	}
+
+	logrus.Infof("all volumes are detached")
+	logrus.Infof("update Longhorn settings")
+	// push LH setting
+	if err := h.updateLonghornStorageNetwork(setting.Value); err != nil {
+		logrus.Warnf("Update Longhorn setting error %v", err)
+	}
+	if _, err := h.setConfiguredCondition(setting, false, ReasonInProgress, MsgUpdateLonghornSetting); err != nil {
+		logrus.Warnf("update status error %v", err)
+	}
+	h.settingsController.EnqueueAfter(setting.Name, 5*time.Second)
+
+	return nil, nil
```

Comment:
Return setting.

Reply:
I'm not sure the reason to return setting here.
If we handle this setting correctly, there is no any other controller should handle this setting again

Reply:
Ok, let's leave it `nil` here. Thanks